### ISSUE TITLE
Namespaces on overrode classes

### ISSUE THREAD

**Issue by NordineA** [PrestaShop/PrestaShop#30112]
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md)
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add screenshots

- Namespaces stays inside the overrode front controller when the module is uninstalled.

modules/mymodule/override/controllers/front/ProductController.php
```
<?php

use PrestaShop\PrestaShop\Adapter\Image\ImageRetriever;
use PrestaShop\PrestaShop\Adapter\Presenter\AbstractLazyArray;
use PrestaShop\PrestaShop\Adapter\Presenter\Product\ProductListingPresenter;
use PrestaShop\PrestaShop\Adapter\Product\PriceFormatter;
use PrestaShop\PrestaShop\Adapter\Product\ProductColorsRetriever;
use PrestaShop\PrestaShop\Core\Addon\Module\ModuleManagerBuilder;
use PrestaShop\PrestaShop\Core\Product\ProductExtraContentFinder;
use PrestaShop\PrestaShop\Core\Product\ProductInterface;

class ProductController extends ProductControllerCore {

    public function initContent() {
        parent::initContent();
    }
}
```

override/controllers/front/ProductController.php

```
<?php
use PrestaShop\PrestaShop\Adapter\Image\ImageRetriever;
use PrestaShop\PrestaShop\Adapter\Presenter\AbstractLazyArray;
use PrestaShop\PrestaShop\Adapter\Presenter\Product\ProductListingPresenter;
use PrestaShop\PrestaShop\Adapter\Product\PriceFormatter;
use PrestaShop\PrestaShop\Adapter\Product\ProductColorsRetriever;
use PrestaShop\PrestaShop\Core\Addon\Module\ModuleManagerBuilder;
use PrestaShop\PrestaShop\Core\Product\ProductExtraContentFinder;
use PrestaShop\PrestaShop\Core\Product\ProductInterface;
class ProductController extends ProductControllerCore {
        parent::initContent();
    }
    * module: mymodule
    * date: 2022-10-25 15:16:51
    * version: 0.1
    */
    public function initContent() {
}
```

### Expected behavior

override/controllers/front/ProductController.php

```
<?php
use PrestaShop\PrestaShop\Adapter\Image\ImageRetriever;
use PrestaShop\PrestaShop\Adapter\Presenter\AbstractLazyArray;
use PrestaShop\PrestaShop\Adapter\Presenter\Product\ProductListingPresenter;
use PrestaShop\PrestaShop\Adapter\Product\PriceFormatter;
use PrestaShop\PrestaShop\Adapter\Product\ProductColorsRetriever;
use PrestaShop\PrestaShop\Core\Addon\Module\ModuleManagerBuilder;
use PrestaShop\PrestaShop\Core\Product\ProductExtraContentFinder;
use PrestaShop\PrestaShop\Core\Product\ProductInterface;
class ProductController extends ProductControllerCore {

    * module: mymodule
    * date: 2022-10-25 15:16:51
    * version: 0.1
    */
    public function initContent() {
         parent::initContent();
    }
}

```

### Steps to reproduce

1. Create a new module MyModule
2. Override the ProductController inside the module in module/mymodule/override/controllers/front/ProductController.php
3. Add the following code inside the module/mymodule/override/controllers/front/ProductController.php
```
<?php
use PrestaShop\PrestaShop\Adapter\Image\ImageRetriever;
use PrestaShop\PrestaShop\Adapter\Presenter\AbstractLazyArray;
use PrestaShop\PrestaShop\Adapter\Presenter\Product\ProductListingPresenter;
use PrestaShop\PrestaShop\Adapter\Product\PriceFormatter;
use PrestaShop\PrestaShop\Adapter\Product\ProductColorsRetriever;
use PrestaShop\PrestaShop\Core\Addon\Module\ModuleManagerBuilder;
use PrestaShop\PrestaShop\Core\Product\ProductExtraContentFinder;
use PrestaShop\PrestaShop\Core\Product\ProductInterface;
class ProductController extends ProductControllerCore {

    public function initContent() {
        parent::initContent();
    }
}
```
4. Install the module
5. Reinitialize the module

You should have an error telling you that you are not able to uninstall the module due to the issue encountered in the overrode ProductController file in override/controllers/front/ProductController.php

### PrestaShop version(s) where the bug happened

1.7.5.1

### PHP version(s) where the bug happened

7.2

**Comment by hibatallahAouadni** [comment]
Hello @NordineA 

Could you please provide a test module to help me reproducing the issue :pray: 
Waiting for your feedback.

Thanks!

**Comment by cdue** [comment]
Hi there,

I think I have the same issue (or something similar) on PS 1.7.7.2.

I have created a module that needs to override `PasswordController.php`, and that override contains a few `use` statements.
When first creating that override, everything worked fine.
But then I had to add one more `use` statement, so I created an upgrade script containing the following line to force a refresh of the existing module overrides :
```
return $module->uninstallOverrides() && $module->installOverrides();
```
But still, my new `use` statement was not added to the generated override class.

I did a little more digging and discovered that:
- If I manually delete the generated override class, then upgrade (or deactivate/active) the module, the newly generated override class contains the new `use` statement
- adding a new `use` statement to a class that doesn't contain any `use` statement yet, works fine.
- uninstalling the overrides from a module upgrade script fully removes an override class if it doesn't contain any `use` statement (and I suppose if there is no other module also adding methods to that override)
- uninstalling the overrides from a module upgrade script only clears the inside of an override class if it contains at least one `use` statement --> this is where my problem begins...

So I came up with a crappy working solution for now...: I programmatically delete the file from an upgrade script before reinstalling the overrides. But I know that it can break other modules which could also add their custom methods to the same override...

just in case anyone needs a quick and dirty fix before someone finds the real one, here is what my upgrade script look like:

```
<?php

if (!defined('_PS_VERSION_')) {
    exit;
}

function upgrade_module_1_1_5($module): bool
{
    $overridePath = _PS_ROOT_DIR_ . '/override/controllers/front/PasswordController.php';

    if (is_file($overridePath)) {
        if(!unlink($overridePath)) {
            return false;
        }
    }

    return $module->uninstallOverrides() && $module->installOverrides();
}

```

I tried to find what to fix in the `Module.php` class, `addOverride` method, but I don't really get how the `use` statements are copied from the source class to the generated one.

I hope this can point someone in the right direction.
Keep up the good work 😉

Cheers

**Comment by kpodemski** [comment]
@cdue I think that instead of `unlink` you can do `$module->removeOverride('PasswordController')` :)

**Comment by cdue** [comment]
> @cdue I think that instead of `unlink` you can do `$module->removeOverride('PasswordController')` :)

Hi @kpodemski,

Thanks for the suggestion but this doesn't work.
In fact, the `removeOverride` method is called by `uninstallOverrides` which is not doing any good with `use` statements.
The PHP Doc of `removeOverride` says: "Remove all methods in a module override from the override class."
So I guess it only works on methods, nothing else.

The problem here is when you first add a `use` statement, pointing to a namespace of your own module (not a Prestashop class), installing the override using `installOverrides` works well, but then if you need to add one more `use` statement in a future version of your module, only methods will be uninstalled and installed in the generated override, `use` statements will remain as they were at first, which will obviously result in a "missing use statement" exception each time the override is called.

For now, I see no other way than unlinking the file (which is bad, I know 😞), but if you have any other suggestion, I'll give it a try 😉.

**Comment by AureRita** [comment]
hi @NordineA ,

we keep waiting for an example module 

waiting for your feedback

**Comment by NordineA** [comment]
Hi,

Here's the example module.

[mymodule.zip](https://github.com/PrestaShop/PrestaShop/files/10013536/mymodule.zip)

**Comment by AureRita** [comment]
Hi @NordineA 

I reproduced your issue on 1.7.7.8 and 1.7.8.8 but not on 8.0.0 so this seems to be fixed on 8.0.0. 

So I"ll closed it at fixed
